File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/list/wordsearch/WordSearchResultYearMonthList.java
package com.websarva.wings.android.zuboradiary.ui.list.wordsearch;

import androidx.annotation.NonNull;

import com.websarva.wings.android.zuboradiary.ui.list.DiaryYearMonthListBaseAdapter.ViewType;

import java.time.LocalDate;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public class WordSearchResultYearMonthList {

    private final List<WordSearchResultYearMonthListItem> wordSearchResultYearMonthListItemList;

    WordSearchResultYearMonthList(WordSearchResultDayList wordSearchResultDayList, boolean needsNoDiaryMessage) {
        Objects.requireNonNull(wordSearchResultDayList);
        if (wordSearchResultDayList.getWordSearchResultDayListItemList().isEmpty()) throw new IllegalArgumentException();

        List<WordSearchResultYearMonthListItem> WordSearchResultYearMonthListItemList = createWordSearchResultYearMonthListItem(wordSearchResultDayList);
        addLastItem(WordSearchResultYearMonthListItemList, needsNoDiaryMessage);
        this.wordSearchResultYearMonthListItemList = Collections.unmodifiableList(WordSearchResultYearMonthListItemList);
    }

    WordSearchResultYearMonthList(List<WordSearchResultYearMonthListItem> itemList, boolean needsNoDiaryMessage) {
        Objects.requireNonNull(itemList);
        if (itemList.isEmpty()) throw new IllegalArgumentException();
        itemList.stream().forEach(Objects::requireNonNull);

        List<WordSearchResultYearMonthListItem> _itemList = new ArrayList<>(itemList);
        addLastItem(_itemList, needsNoDiaryMessage);
        this.wordSearchResultYearMonthListItemList = Collections.unmodifiableList(_itemList);
    }

    /**
     * true:日記なしメッセージのみのリスト作成<br>
     * false:ProgressIndicatorのみのリスト作成
     * */
    WordSearchResultYearMonthList(boolean needsNoDiaryMessage) {
        List<WordSearchResultYearMonthListItem> itemList = new ArrayList<>();
        addLastItem(itemList, needsNoDiaryMessage);
        this.wordSearchResultYearMonthListItemList = Collections.unmodifiableList(itemList);
    }

    WordSearchResultYearMonthList() {
        List<WordSearchResultYearMonthListItem> itemList = new ArrayList<>();
        this.wordSearchResultYearMonthListItemList = Collections.unmodifiableList(itemList);
    }

    @NonNull
    private List<WordSearchResultYearMonthListItem> createWordSearchResultYearMonthListItem(WordSearchResultDayList wordSearchResultDayList) {
        Objects.requireNonNull(wordSearchResultDayList);
        if (wordSearchResultDayList.getWordSearchResultDayListItemList().isEmpty()) throw new IllegalArgumentException();

        List<WordSearchResultDayListItem> sortingDayItemList= new ArrayList<>();
        List<WordSearchResultYearMonthListItem> resultYearMonthListItemList = new ArrayList<>();
        WordSearchResultYearMonthListItem resultYearMonthListItem;
        YearMonth sortingYearMonth = null;

        List<WordSearchResultDayListItem> resultDayListItemList = wordSearchResultDayList.getWordSearchResultDayListItemList();
        for (WordSearchResultDayListItem day: resultDayListItemList) {
            LocalDate date = day.getDate();
            YearMonth yearMonth = YearMonth.of(date.getYear(), date.getMonth());

            if (sortingYearMonth != null && !yearMonth.equals(sortingYearMonth)) {
                WordSearchResultDayList sortedWordSearchResultDayList = new WordSearchResultDayList(sortingDayItemList);
                resultYearMonthListItem =
                        new WordSearchResultYearMonthListItem(sortingYearMonth, sortedWordSearchResultDayList);
                resultYearMonthListItemList.add(resultYearMonthListItem);
                sortingDayItemList= new ArrayList<>();
            }
            sortingDayItemList.add(day);
            sortingYearMonth = yearMonth;
        }

        WordSearchResultDayList sortedWordSearchResultDayList = new WordSearchResultDayList(sortingDayItemList);
        resultYearMonthListItem =
                new WordSearchResultYearMonthListItem(sortingYearMonth, sortedWordSearchResultDayList);
        resultYearMonthListItemList.add(resultYearMonthListItem);
        return resultYearMonthListItemList;
    }

    private void addLastItem(List<WordSearchResultYearMonthListItem> itemList, boolean needsNoDiaryMessage) {
        Objects.requireNonNull(itemList);
        itemList.stream().forEach(Objects::requireNonNull);

        itemList.removeIf(x -> !x.isDiaryViewType());
        if (needsNoDiaryMessage) {
            addLastItemNoDiaryMessage(itemList);
        } else {
            addLastItemProgressIndicator(itemList);
        }
    }

    private void addLastItemProgressIndicator(List<WordSearchResultYearMonthListItem> itemList) {
        Objects.requireNonNull(itemList);
        itemList.stream().forEach(Objects::requireNonNull);

        itemList.add(new WordSearchResultYearMonthListItem(ViewType.PROGRESS_INDICATOR));
    }

    private void addLastItemNoDiaryMessage(List<WordSearchResultYearMonthListItem> itemList) {
        Objects.requireNonNull(itemList);
        itemList.stream().forEach(Objects::requireNonNull);

        itemList.add(new WordSearchResultYearMonthListItem(ViewType.NO_DIARY_MESSAGE));
    }

    public int countDiaries() {
        int count = 0;
        for (WordSearchResultYearMonthListItem item: wordSearchResultYearMonthListItemList) {
            if (item.getViewType().equals(ViewType.DIARY)) {
                count += item.getWordSearchResultDayList().countDiaries();
            }
        }
        return count;
    }

    WordSearchResultYearMonthList combineDiaryLists(
            WordSearchResultYearMonthList additionList, boolean needsNoDiaryMessage) {
        Objects.requireNonNull(additionList);
        if (additionList.getWordSearchResultYearMonthListItemList().isEmpty()) throw new IllegalArgumentException();

        List<WordSearchResultYearMonthListItem> originalItemList = new ArrayList<>(this.wordSearchResultYearMonthListItemList);
        List<WordSearchResultYearMonthListItem> additionItemList = new ArrayList<>(additionList.wordSearchResultYearMonthListItemList);

        // List最終アイテム(日記以外
        originalItemList.removeIf(x -> !x.isDiaryViewType());
        additionItemList.removeIf(x -> !x.isDiaryViewType());

        // 元リスト最終アイテムの年月取得
        int originalListLastItemPosition = originalItemList.size() - 1;
        WordSearchResultYearMonthListItem originalListLastItem = originalItemList.get(originalListLastItemPosition);
        YearMonth originalListLastItemYearMonth = originalListLastItem.getYearMonth();

        // 追加リスト先頭アイテムの年月取得
        WordSearchResultYearMonthListItem additionListFirstItem = additionList.getWordSearchResultYearMonthListItemList().get(0);
        YearMonth additionListFirstItemYearMonth = additionListFirstItem.getYearMonth();

        // 元リストに追加リストの年月が含まれていたらアイテムを足し込む
        if (originalListLastItemYearMonth.equals(additionListFirstItemYearMonth)) {
            WordSearchResultDayList originalLastWordSearchResultDayList =
                    originalItemList.get(originalListLastItemPosition).getWordSearchResultDayList();
            WordSearchResultDayList additionWordSearchResultDayList = additionListFirstItem.getWordSearchResultDayList();
            WordSearchResultDayList combinedWordSearchResultDayList =
                    originalLastWordSearchResultDayList.combineDiaryDayLists(additionWordSearchResultDayList);
            WordSearchResultYearMonthListItem combinedResultYearMonthListItem
                    = new WordSearchResultYearMonthListItem(originalListLastItemYearMonth, combinedWordSearchResultDayList);
            originalItemList.remove(originalListLastItemPosition);
            originalItemList.add(combinedResultYearMonthListItem);
            additionItemList.remove(0);
        }

        List<WordSearchResultYearMonthListItem> resultItemList = new ArrayList<>(originalItemList);
        resultItemList.addAll(additionItemList);

        return new WordSearchResultYearMonthList(resultItemList, needsNoDiaryMessage);
    }

    public List<WordSearchResultYearMonthListItem> getWordSearchResultYearMonthListItemList() {
        return wordSearchResultYearMonthListItemList;
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/adapter/diary/wordsearchresult/WordSearchResultYearMonthList.kt
package com.websarva.wings.android.zuboradiary.ui.adapter.diary.wordsearchresult

import com.websarva.wings.android.zuboradiary.ui.adapter.diary.DiaryYearMonthListBaseAdapter
import com.websarva.wings.android.zuboradiary.ui.adapter.diary.DiaryYearMonthListBaseItem
import java.time.YearMonth

internal class WordSearchResultYearMonthList {

    val itemList: List<WordSearchResultYearMonthListItem>

    constructor(
        wordSearchResultDayList: WordSearchResultDayList,
        needsNoDiaryMessage: Boolean
    ) {
        require(wordSearchResultDayList.itemList.isNotEmpty())

        val itemList =
            createWordSearchResultYearMonthListItem(wordSearchResultDayList)
        this.itemList = addLastItem(itemList, needsNoDiaryMessage)
    }

    constructor(
        itemList: List<WordSearchResultYearMonthListItem>,
        needsNoDiaryMessage: Boolean
    ) {
        require(itemList.isNotEmpty())

        this.itemList = addLastItem(itemList, needsNoDiaryMessage)
    }

    /**
     * true:日記なしメッセージのみのリスト作成<br></br>
     * false:ProgressIndicatorのみのリスト作成
     */
    constructor(needsNoDiaryMessage: Boolean) {
        val emptyList: List<WordSearchResultYearMonthListItem> = ArrayList()
        this.itemList = addLastItem(emptyList, needsNoDiaryMessage)
    }

    constructor() {
        this.itemList = ArrayList()
    }

    private fun createWordSearchResultYearMonthListItem(
        wordSearchResultDayList: WordSearchResultDayList
    ): List<WordSearchResultYearMonthListItem> {
        require(wordSearchResultDayList.itemList.isNotEmpty())

        var sortingDayItemList: MutableList<WordSearchResultDayListItem> = ArrayList()
        val resultYearMonthListItemList: MutableList<WordSearchResultYearMonthListItem> = ArrayList()
        var resultYearMonthListItem: WordSearchResultYearMonthListItem
        var sortingYearMonth: YearMonth? = null

        // DayListItemを対象YearMonthListItem毎に仕分け
        val resultDayListItemList = wordSearchResultDayList.itemList
        for (day in resultDayListItemList) {
            val date = day.date
            val yearMonth = YearMonth.of(date.year, date.month)

            if (sortingYearMonth != null && yearMonth != sortingYearMonth) {
                val sortedWordSearchResultDayList = WordSearchResultDayList(sortingDayItemList)
                resultYearMonthListItem =
                    WordSearchResultYearMonthListItem(
                        sortingYearMonth,
                        sortedWordSearchResultDayList
                    )
                resultYearMonthListItemList.add(resultYearMonthListItem)
                sortingDayItemList = ArrayList()
            }
            sortingDayItemList.add(day)
            sortingYearMonth = yearMonth
        }

        // 最後尾YearMonthListItemの追加処理
        val sortedWordSearchResultDayList = WordSearchResultDayList(sortingDayItemList)
        resultYearMonthListItem =
            WordSearchResultYearMonthListItem(
                checkNotNull(sortingYearMonth),
                sortedWordSearchResultDayList
            )
        resultYearMonthListItemList.add(resultYearMonthListItem)

        return resultYearMonthListItemList
    }

    private fun addLastItem(
        itemList: List<WordSearchResultYearMonthListItem>,
        needsNoDiaryMessage: Boolean
    ): List<WordSearchResultYearMonthListItem> {
        val mutableItemList = itemList.toMutableList()
        mutableItemList.removeIf(DiaryYearMonthListBaseItem::isNotDiaryViewType)

        return if (needsNoDiaryMessage) {
            addLastItemNoDiaryMessage(mutableItemList)
        } else {
            addLastItemProgressIndicator(mutableItemList)
        }
    }

    private fun addLastItemProgressIndicator(
        itemList: List<WordSearchResultYearMonthListItem>
    ): List<WordSearchResultYearMonthListItem> {

        return itemList +
                WordSearchResultYearMonthListItem(
                    DiaryYearMonthListBaseAdapter.ViewType.PROGRESS_INDICATOR
                )
    }

    private fun addLastItemNoDiaryMessage(
        itemList: List<WordSearchResultYearMonthListItem>
    ): List<WordSearchResultYearMonthListItem> {
        return itemList +
                WordSearchResultYearMonthListItem(
                    DiaryYearMonthListBaseAdapter.ViewType.NO_DIARY_MESSAGE
                )
    }

    fun countDiaries(): Int {
        var count = 0
        for (item in itemList) {
            if (item.viewType == DiaryYearMonthListBaseAdapter.ViewType.DIARY) {
                count += item.wordSearchResultDayList.countDiaries()
            }
        }
        return count
    }

    fun combineDiaryLists(
        additionList: WordSearchResultYearMonthList, needsNoDiaryMessage: Boolean
    ): WordSearchResultYearMonthList {
        require(additionList.itemList.isNotEmpty())

        val originalItemList: MutableList<WordSearchResultYearMonthListItem> =
            itemList.toMutableList()
        val additionItemList: MutableList<WordSearchResultYearMonthListItem> =
            additionList.itemList.toMutableList()

        // List最終アイテム(日記以外
        originalItemList.removeIf(DiaryYearMonthListBaseItem::isNotDiaryViewType)
        additionItemList.removeIf(DiaryYearMonthListBaseItem::isNotDiaryViewType)

        // 元リスト最終アイテムの年月取得
        val originalListLastItemPosition = originalItemList.size - 1
        val originalListLastItem = originalItemList[originalListLastItemPosition]
        val originalListLastItemYearMonth = originalListLastItem.yearMonth

        // 追加リスト先頭アイテムの年月取得
        val additionListFirstItem = additionList.itemList[0]
        val additionListFirstItemYearMonth = additionListFirstItem.yearMonth

        // 元リストに追加リストの年月が含まれていたらアイテムを足し込む
        if (originalListLastItemYearMonth == additionListFirstItemYearMonth) {
            val originalLastWordSearchResultDayList =
                originalItemList[originalListLastItemPosition].wordSearchResultDayList
            val additionWordSearchResultDayList = additionListFirstItem.wordSearchResultDayList
            val combinedWordSearchResultDayList =
                originalLastWordSearchResultDayList.combineDiaryDayLists(
                    additionWordSearchResultDayList
                )
            val combinedResultYearMonthListItem = WordSearchResultYearMonthListItem(
                originalListLastItemYearMonth,
                combinedWordSearchResultDayList
            )
            originalItemList.removeAt(originalListLastItemPosition)
            originalItemList.add(combinedResultYearMonthListItem)
            additionItemList.removeAt(0)
        }

        val resultItemList = originalItemList + additionItemList
        return WordSearchResultYearMonthList(resultItemList, needsNoDiaryMessage)
    }
}
